api/src/utils.ts

```diff
@@ -0,0 +1,64 @@
+import { MiddlewareHandler } from "hono";
+
+export const entries = <T extends object>(o: T) => {
+	return Object.entries(o) as [keyof T, T[keyof T]][];
+};
+export const filterObject = <T extends object>(
+	o: T,
+	predicate: (o: [keyof T, T[keyof T]]) => boolean,
+) => {
+	return Object.fromEntries(entries(o).filter(([key, value]) => predicate([key, value])));
+};
+
+export const mapObject = <T extends object, R>(
+	o: T,
+	func: (o: [keyof T, T[keyof T]]) => [keyof T, R],
+): Record<keyof T, R> => {
+	return Object.fromEntries(entries(o).map(([k, v]) => func([k, v]))) as Record<keyof T, R>;
+};
+
+export const languagesFilter =
+	(nested: boolean): MiddlewareHandler =>
+	async (c, next) => {
+		const languages = c.req.query("lang")?.split(",") ?? ["en"];
+		await next();
+
+		if (languages.length === 1 && languages[0] === "*") return;
+
+		const body = (await c.res.json()) as any;
+		if (nested) {
+			c.res = new Response(
+				JSON.stringify(
+					Object.fromEntries(
+						Object.entries(body)
+							.filter(
+								(e): e is [string, { translations: any }] =>
+									typeof e[1] === "object" && !!e[1] && "translations" in e[1],
+							)
+							.map(([k, v]) => [
+								k,
+								{
+									...v,
+									translations: filterObject(v["translations"], ([k]) =>
```

Comment:
Opinion: The API should throw an error if the user requests a language that doesn't exist. Otherwise the translation field comes back empty. This isn't a bad behavior necessarily, but it's an unexpected one.

Alternatively, we could return both the expected object and a status object, where the status object indicates the success (success, partial success [error object e.g. one or more languages not found], failure [error object e.g. malformed query]).

Thoughts?

Reply:
This isn't necessarily the case, since users can request multiple languages at once, some of which may or may not be valid: `/words?lang=en,xx,de,random_stuff_that_isnt_valid`, and the idea behind this was that if you ask for a mixture of valid and invalid languages, it wouldn't fail.